lib/mwUtil.js

```diff
@@ -386,4 +386,47 @@ mwUtil.isSelfRedirect = (req, location) => {
     return true;
 };
 
+/**
+ * Traverses through a request body and replaces the given keys with
+ * the content fetched using the 'fetch' callback
+ *
+ * @param response
+ * @param fetch
+ */
+mwUtil.hydrateResponse = (response, fetch) => {
+    const requests = {};
+    const setters = [];
+
+    function _traverse(node, removeFromParent) {
+        function requestResource(resource) {
+            requests[resource] = requests[resource] || fetch(resource);
+            setters.push((content) => {
+                if (content[resource]) {
+                    Object.assign(node, content[resource]);
+                } else if (removeFromParent) {
+                    removeFromParent();
+                }
+                delete node.$merge;
+            });
+        }
+
+        if (Array.isArray(node)) {
+            for (let i = 0; i < node.length; i++) {
+                _traverse(node[i], () => node.splice(i, 1));
```

Comment:
This doesn't traverse the whole array, it skips every second element, because `splice()` removes the current element, and then `i` is increased:

```js
> let a = [1,2,3,4,5];
> for(let i = 0; i < a.length; i++) { console.log(i, a[i]); a.splice(i, 1); }
0 1
1 3
2 5
[ 5 ]
> a
[ 2, 4 ]
```